File: app/src/main/java/danb/speedrunbrowser/api/objects/Game.java
package danb.speedrunbrowser.api.objects;

import android.widget.TextView;

import java.io.Serializable;
import java.net.URL;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

public class Game implements Serializable, SearchResultItem {
    public String id;
    public HashMap<String, String> names;
    public String abbreviation;
    public String weblink;

    public int released;
    public String releaseDate;
    public GameRuleset ruleset;
    public boolean romhack;
    public List<String> gametypes;
    public List<Platform> platforms;
    public List<Region> regions;
    public List<String> genres;
    public List<String> engines;
    public List<String> developers;
    public List<String> publishers;
    public HashMap<String, String> moderators;
    public Date created;

    public List<Category> categories;
    public List<Level> levels;

    public GameAssets assets;

    public String getName() {
        String n = names.get("international");
        return n != null ? n : "? Unknown Name ?";
    }

    public String getTypeName() {
        return "game";
    }

    public URL getIconUrl() {
        return assets.coverLarge.uri;
    }

    public void applyTextView(TextView tv) {
        tv.setText(getName());
    }

    public boolean shouldShowPlatformFilter() {
        return platforms != null && platforms.size() > 1;
    }

    public boolean shouldShowRegionFilter() {
        return regions != null && regions.size() > 1;
    }
}

File: app/src/main/java/danb/speedrunbrowser/api/objects/Game.kt
package danb.speedrunbrowser.api.objects

import android.widget.TextView
import java.lang.reflect.Type
import java.io.Serializable
import java.net.URL
import java.util.*

import com.google.gson.*

data class GameMaker(
        val id: String,
        val name: String
) : Serializable {
    class JsonConverter : JsonDeserializer<GameMaker> {

        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): GameMaker {

            if (json.isJsonObject) {
                val obj = json.asJsonObject
                return GameMaker(
                        id = obj.get("id").asString,
                        name = obj.get("name").asString
                )
            }
            else {
                return GameMaker(
                        id = json.asString,
                        name = ""
                )
            }
        }
    }
}

data class Game(
    val id: String,
    val names: Map<String, String>? = mapOf(),
    val abbreviation: String = "",
    val weblink: String = "",

    val released: Int = 0,
    val releaseDate: String? = null,
    val ruleset: GameRuleset? = null,
    val romhack: Boolean = false,
    val gametypes: List<String>? = null,
    val platforms: List<Platform>? = null,
    val regions: List<Region>? = null,
    val genres: List<Genre>? = null,
    val engines: List<String>? = null,
    val developers: List<GameMaker>? = null,
    val publishers: List<GameMaker>? = null,
    val moderators: HashMap<String, String>? = null,
    val created: Date? = null,

    val categories: List<Category>? = null,
    val levels: List<Level>? = null,

    val assets: GameAssets = GameAssets()
) : Serializable, SearchResultItem {

    override val resolvedName: String
    get() = names?.get("international") ?: "? Unknown Name ?"

    override val type: String
            get() = "game"

    override val iconUrl: URL?
        get() = assets.coverLarge?.uri ?: assets.icon!!.uri

    override fun applyTextView(tv: TextView) {
        tv.text = resolvedName
    }

    fun shouldShowPlatformFilter(): Boolean {
        return platforms != null && platforms.size > 1
    }

    fun shouldShowRegionFilter(): Boolean {
        return regions != null && regions.size > 1
    }
}
